docs/dev_guide/code_coverage_guide.md

```diff
@@ -0,0 +1,105 @@
+---
+layout: default
+title: Code Coverage Guide
+parent: Developer Guides
+permalink: /docs/dev_guide/code_coverage_guide
+---
+# Code Coverage Guide
+
+The Venice repository has two code coverage check on GitHub Action, submodule-level coverage verification
+and new-commit coverage verification, which aim to improve the testability and code quality overall. The
+GitHub Action fails the Pull Request if the coverage check doesn't meet the requirements. This guide provides the details
+on how these reports are generated and how to debug them.
+
+## Module structure
+
+Venice has 4 modules, `all-modules`, `clients`, `internal` and `services`. Each module has its own submodules, except `all-modules`.
+
+#### Clients
+
+`da-vinci-client`, `venice-admin-tool`, `venice-client`, `venice-push-job`, `venice-samza`, `venice-thin-client`
+
+#### Internal
+`alpini`, `venice-avro-compatibility-test`, `venice-client-common`, `venice-common`, `venice-consumer`, 
+`venice-jdk-compatibility-test`, `venice-test-common`
+
+#### Services
+
+`venice-router`, `venice-controller`, `venice-server`
+
+## Submodule-level coverage verification
+
+### Description
+This will check the overall code coverage at the submodule level.
+Jacoco generates the report based off the current submodule, perform the coverage verification and fail the build if the
+coverage is below the threshold. The threshold, targets at the branch coverage, is defined in the `build.gradle` of each 
+submodule independently.
+
+### Example commands
+```Bash
+# Template
+./gradlew :<module name>:<submodule name>:jacocoTestCoverageVerification 
+
+./gradlew :clients:venice-push-job:jacocoTestCoverageVerification 
+```
+
+The report will be located at `<module name>/<submodule name>/build/reports/jacoco/test/index.html`.
+
+
+### Debugging notes
+
+#### Run commands against module doesn't generate the report
+
+Due to the current project setup, running the commands, e.g. `./gradlew :clients:jacocoTestCoverageVerification`, doesn't
+execute the unit tests thus no jacoco report will be generated. Please be sure to run the commands against the submodule.
+
+## New-commit coverage verification
+
+### Description
+This will check the overall code coverage at the commit-level.
+DiffCoverage, which is an extension of Jacoco, gathers the diff files by comparing the local branch and remote upstream 
+main branch, and leverage the Jacoco report, to re-generate a new report only for these newly added lines/files. Similarly,
+it performs the coverage verification and fail the build if the coverage is below the threshold. The threshold is defined 
+at 60% for branch coverage.
```

Comment:
I wonder if a fixed number makes sense for this... is there any other option on how to set this? (E.g. based on module or based on what it was previously in the main branch?)